GCC 11 doesn't support ranges properly

ennorehling opened the pull request:
On Ubuntu 22.04, the default GNU C compiler version is gcc-11. It doesn't understand the code in string_filters.cpp, because the ranges library wasn't fully implemented at the time.

It's possible to achieve the same result with a slightly less efficient implementation, which I'm only enabling if required.

It would of course be nicer to detect the compiler and set -DWITHOUT_VIEWS in the Makefile, not in string_filters.hpp, but I don't know how to do that.

jt-traub commented:
As I said on the game channel, I would prefer to support only libc++20 or newer robustly, which is pretty much g++12 or newer.  I appreciate that this would make it more available to a number of people (and yes, the default g++ on 22.04 is gcc-11, but gcc-12 is available and officially supported and (supposedly since I haven't tested it) fully supports the ranges library).  If that does work, I would prefer not to support older version than that at this time.

jt-traub commented on string_parser.hpp:
```diff
@@ -55,7 +54,7 @@ std::istream& operator>>(std::istream& is, string_parser& parser);
 class string_parser {
 public:
-    constexpr string_parser() noexcept : data_{}, pos_{0} {}
+    string_parser() noexcept : data_{}, pos_{0} {}
```
Am curious on the reason for removal of the constexpr here?

ennorehling replied:
I had some compilation problems with them, don't remember from which compiler. Removing the constexpr made the code build, and I never thought about it again. I'm not incredibly versed in the correct use of this keyword, but if you want, I'll find out what my issue was, and if there's a better solution.

jt-traub replied:
If so, we might need to figure something out.

jt-traub commented on string_filters.hpp:
```diff
@@ -124,6 +133,23 @@ inline constexpr lowercase_t lowercase{};
+            }
+            s += w | capitalize;
+        }
+        return str;
```
in fact, I believe this should return strings::join(words, "_");

I would need to verify this using that version of gcc to be sure this is correct however.

ennorehling replied:
I think you're right, this should return s. These filter functions could have their own unit tests, I suppose?

jt-traub commented on string_filters.hpp:
```diff
@@ -124,6 +133,23 @@ inline constexpr lowercase_t lowercase{};
+        std::istringstream f(str);
+        std::string s;
+        while (std::getline(f, s, ' ')) {
+            words.push_back(s);
```
To go with the return suggestion I made below, this should probably be words.push_back(s | capitalize)

ennorehling commented:
I added a test and fixed the non-ranges canonicalize implementation. Also decided to rename the preprocessor flag that enables it, and removed the compiler version checks, to be set by Makefile or CMake instead.

jt-traub commented on string_filters.hpp:
```diff
@@ -124,6 +128,16 @@ inline constexpr lowercase_t lowercase{};
  */
 struct canonicalize_t {
     std::string process(const std::string& str) const {
+#ifdef CPP20_RANGES_ARE_BROKEN
+        /** gcc 11 doesn't like the range solution, ergo this hack */
+        std::vector<std::string> words;
+        std::istringstream f(str);
+        std::string s;
+        while (std::getline(f, s, ' ')) {
+            words.push_back(s | capitalize);
+        }
+        return strings::join(words, "_");
+#else
```
This could actually be tightened up further I believe.
```suggestion
        std::vector<std::string> processed_words;
#ifdef CPP20_RANGES_ARE_BROKEN
        /** gcc 11 doesn't like the range solution, ergo this hack */
        std::istringstream f(str);
        std::string s;
        while (std::getline(f, s, ' ')) {
            processed_words.push_back(s | capitalize);
        }
#else
```

jt-traub replied:
Then in the else block, remove the definition of processed_words, and the return, and then

jt-traub commented on string_filters.hpp:
```diff
@@ -136,6 +150,7 @@ struct canonicalize_t {
         std::ranges::copy(parts_view, std::back_inserter(processed_words));
         // Join the processed words with an underscore
         return strings::join(processed_words, "_");
+#endif
```
```suggestion
#endif
        return strings::join(processed_words, "_");
```

Doing it this way just changes the way processed_words are built and keeps the rest of the logic uniform, which is clearer intent imho.

jt-traub reviewed (commented):
Thanks for adding those unit tests.  If we can get this branch rebased onto current master to resolve the merge conflicts, then I can approve it.
